autopush/web/validation.py

```diff
@@ -0,0 +1,166 @@
+    in a separate thread for a cyclone request handler.
+
+    """
+    log = Logger()
+
+    def __init__(self, schema):
+        self.schema = schema
+
+    def _validate_request(self, request_handler):
+        """Validates a schema_class against a cyclone request"""
+        data = {
+            "headers": request_handler.request.headers,
+            "body": request_handler.request.body,
+            "path_args": request_handler.path_args,
+            "path_kwargs": request_handler.path_kwargs,
+            "arguments": request_handler.request.arguments,
+        }
+        self.schema.context["settings"] = request_handler.ap_settings
+        self.schema.context["log"] = self.log
+        return self.schema.load(data)
+
+    def _call_func(self, result, func, request_handler, *args, **kwargs):
+        output, errors = result
+        if errors:
+            return request_handler._write_validation_err(errors)
+        request_handler.valid_input = output
+        return func(request_handler, *args, **kwargs)
+
+    def _decorator(self, func):
+        @wraps(func)
+        def wrapper(request_handler, *args, **kwargs):
+            # Wrap the handler in @cyclone.web.synchronous
+            request_handler._auto_finish = False
+
+            d = deferToThread(self._validate_request, request_handler)
+            d.addErrback(request_handler._overload_err)
+            d.addErrback(request_handler._validation_err)
+            d.addErrback(request_handler._response_err)
+            d.addCallback(self._call_func, func, request_handler, *args,
+                          **kwargs)
+        return wrapper
+
+    @classmethod
+    def validate(cls, schema):
+        """Validate a request schema in a separate thread before calling the
+        request handler
+
+        An alias `threaded_validate` should be used from this module.
+
+        Using `cyclone.web.asynchronous` is not needed as this function
+        will attach equivilant functionality to the method handler. Calling
+        `self.finish()` is needed on decorated handlers.
+
+        .. code-block::
+
+            class MyHandler(cyclone.web.RequestHandler):
+                @threaded_validate(MySchema())
+                def post(self):
+                    ...
+
+        """
+        return cls(schema)._decorator
+
+
+# Alias to the validation classmethod decorator
+threaded_validate = ThreadedValidate.validate
+
+
+class SimplepushSubscriptionSchema(Schema):
+    uaid = fields.UUID(required=True)
+    chid = fields.UUID(required=True)
+
+    @pre_load
+    def extract_subscription(self, d):
+        try:
+            result = self.context["settings"].parse_endpoint(
+                token=d["token"],
+                version=d["api_ver"],
+            )
+        except InvalidTokenException:
+            raise InvalidRequest("invalid token", errno=102)
+        return result
+
+    @validates_schema
+    def validate_uaid_chid(self, d):
```

Comment:
OK...so Marshmallow runs the `@pre_load` decorators first, to build the validation object, then checks and converts the fields, then calls `validates_schema` with the modified object? And calling `schema.load()` kicks off the process and returns the object?